src/libsystemd/sd-event/event-util.c

```diff
@@ -177,3 +177,61 @@ dual_timestamp* event_dual_timestamp_now(sd_event *e, dual_timestamp *ts) {
         assert_se(sd_event_now(e, CLOCK_MONOTONIC, &ts->monotonic) >= 0);
         return ts;
 }
+
+void event_source_unref_many(sd_event_source **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
+                sd_event_source_unref(*v);
+
+        free(array);
+}
+
+static int event_forward_signal_callback(sd_event_source *s, const struct signalfd_siginfo *ssi, void *userdata) {
+        sd_event_source *child = ASSERT_PTR(userdata);
+
+        assert(ssi);
+
+        siginfo_t si = {
+                .si_signo = ssi->ssi_signo,
+                /* This has to be SI_QUEUE as SI_USER and SI_KERNEL are not allowed to be used. */
+                .si_code = SI_QUEUE,
+                .si_errno = ssi->ssi_errno,
+        };
+
+        si.si_pid = ssi->ssi_pid;
+        si.si_uid = ssi->ssi_uid;
+        si.si_int = ssi->ssi_int;
+        si.si_ptr = UINT64_TO_PTR(ssi->ssi_ptr);
+
+        return sd_event_source_send_child_signal(child, ssi->ssi_signo, &si, /* flags = */ 0);
+}
+
+int event_forward_signals(sd_event *e, sd_event_source *child, sd_event_source ***ret_sources, size_t *ret_n_sources) {
+        static const int forward_signals[] = { SIGTERM, SIGINT, SIGQUIT, SIGTSTP, SIGCONT, SIGUSR1, SIGUSR2 };
+        sd_event_source **sources = NULL;
+        size_t n_sources = 0;
+        int r;
+
+        CLEANUP_ARRAY(sources, n_sources, event_source_unref_many);
+
+        assert(e);
+        assert(child);
+        assert(ret_sources);
+        assert(ret_n_sources);
+
+        sources = new0(sd_event_source*, ELEMENTSOF(forward_signals));
+        if (!sources)
+                return -ENOMEM;
+
+        FOREACH_ELEMENT(sig, forward_signals) {
+                r = sd_event_add_signal(e, &sources[n_sources], *sig | SD_EVENT_SIGNAL_PROCMASK, event_forward_signal_callback, child);
```

Comment:
I think it is better to increase the reference counter of `child`, and set destroy callback by `sd_event_source_set_destroy_callback()` to unref the child object.